File: firestorelivedata/src/test/java/com/ptrbrynt/firestorelivedata/FirestoreResourceTest.java
package com.ptrbrynt.firestorelivedata;

import org.junit.Test;

import static org.junit.Assert.assertEquals;
import static org.junit.Assert.assertTrue;

public class FirestoreResourceTest {

    @Test
    public void success() {
        FirestoreResource<String> result = FirestoreResource.Companion.success("hello");
        assertEquals(result.getStatus(), Status.SUCCESS);
        assertEquals(result.getData(), "hello");
        assertEquals(result.getThrowable(), null);
        assertEquals(result.getErrorMessage(), null);
    }

    @Test
    public void error() {
        FirestoreResource<String> result = FirestoreResource.Companion.error(new IllegalStateException("You've been played"));
        assertEquals(result.getStatus(), Status.ERROR);
        assertEquals(result.getData(), null);
        assertTrue(result.getThrowable() instanceof IllegalStateException);
        assertEquals(result.getErrorMessage(), "You've been played");
    }

    @Test
    public void loading() {
        FirestoreResource<String> result = FirestoreResource.Companion.loading();
        assertEquals(result.getStatus(), Status.LOADING);
        assertEquals(result.getData(), null);
        assertEquals(result.getThrowable(), null);
        assertEquals(result.getErrorMessage(), null);
    }

}

File: firestorelivedata/src/test/java/com/ptrbrynt/firestorelivedata/FirestoreResourceTest.kt
package com.ptrbrynt.firestorelivedata

import org.junit.Assert.assertEquals
import org.junit.Assert.assertTrue
import org.junit.Test

/**
 * Tests the [FirestoreResource] class
 */
class FirestoreResourceTest {

    /** Tests the success state of [FirestoreResource], ensuring that data is passed correctly and there is no error. **/
    @Test
    fun success() {
        val result = FirestoreResource.success("hello")
        assertEquals(result.status, Status.SUCCESS)
        assertEquals(result.data, "hello")
        assertEquals(result.throwable, null)
        assertEquals(result.errorMessage, null)
    }

    /** Tests the error state of [FirestoreResource], ensuring there is no data and that the error is passed successfully. **/
    @Test
    fun error() {
        val result = FirestoreResource.error<String>(IllegalStateException("You've been played"))
        assertEquals(result.status, Status.ERROR)
        assertEquals(result.data, null)
        assertTrue(result.throwable is IllegalStateException)
        assertEquals(result.errorMessage, "You've been played")
    }

    /** Tests the loading state of [FirestoreResource], ensuring there is no data or error. **/
    @Test
    fun loading() {
        val result = FirestoreResource.loading<String>()
        assertEquals(result.status, Status.LOADING)
        assertEquals(result.data, null)
        assertEquals(result.throwable, null)
        assertEquals(result.errorMessage, null)
    }

}
